codelists/forms.py

```diff
@@ -68,6 +68,27 @@ def __init__(self, *args, **kwargs):
         self.helper.form_tag = False
         super().__init__(*args, **kwargs)
 
+    def full_clean(self):
+        """
+        Override full_clean() to validate Name is unique to a Project.
+
+        This form doesn't include the `project` field from the Codelist model.
+        However, Codelists are unique based on their Project and Name.  Since
+        this is a ModelForm, the view can pass in a Codelist instance with a
+        Project set.  This is further populated with values from the form
+        before we manually validate its uniqueness.
+        """
+        # Call super()'s full_clean to populate the Codelist instance with
+        # values from the form.
+        super().full_clean()
+
+        # Validate uniqueness of the instance now it's been populated with
+        # values from the form and the Project instance.
+        try:
+            self.instance.validate_unique()
+        except forms.ValidationError as e:
+            self._update_errors(e)
```

Comment:
Do we need this stuff on the `CodelistUpdateForm` too?

Reply:
Yes, extracted to a mixin (after much testing pain 😫) in ffcc49f